src/compiler/checker.ts

```diff
@@ -16196,11 +16196,17 @@ namespace ts {
             const type = tryGetThisTypeAt(node, container);
             if (!type && noImplicitThis) {
                 // With noImplicitThis, functions may not reference 'this' if it has type 'any'
-                error(
+                const diag = error(
                     node,
                     capturedByArrowFunction && container.kind === SyntaxKind.SourceFile ?
                         Diagnostics.The_containing_arrow_function_captures_the_global_value_of_this_which_implicitly_has_type_any :
                         Diagnostics.this_implicitly_has_type_any_because_it_does_not_have_a_type_annotation);
+                if (!isSourceFile(container)) {
+                    const outsideThis = tryGetThisTypeAt(container);
+                    if (outsideThis) {
+                        addRelatedInfo(diag, createDiagnosticForNode(container, Diagnostics.There_is_a_this_type_shadowed_by_this_container));

```

Comment:
It would also be nice if we could catch this outside of `noImplicitThis` scenarios.

Reply:
There's no error to hang a span on then 🤷‍♂️  maybe we could track issuing a suggestion along the lines of what @sandersn did with `noImplicitAny` suggestions?

Reply:
I think we do something similar to unused locals etc as well